src/toil/test/src/promisedRequirementTest.py

```diff
@@ -0,0 +1,157 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+import os, fcntl, time, tempfile, multiprocessing
+import logging
+from toil.job import Job
+from toil.job import PromisedRequirement
+from toil.test import ToilTest
+
+log = logging.getLogger(__name__)
+
+class MaxCoresPromisedRequirementTest(ToilTest):
+    """
+    This test ensures that the PromisedRequirement method properly assigns
+    resource requirements to the single machine batch system.
+    """
+
+    def setUp(self):
+        super(MaxCoresPromisedRequirementTest, self).setUp()
+        logging.basicConfig(level=logging.INFO)
+
+        self.counterPath = writeTempFile('0,0')
+        self.cpu_count = multiprocessing.cpu_count()
+
+    def test_dynamic(self):
+        """
+        Tests using a PromisedRequirement object to allocate the number of cores per function
+        """
+        options = Job.Runner.getDefaultOptions(self._getTestJobStorePath())
+        options.logLevel = "INFO"
+        Job.Runner.startToil(Job.wrapJobFn(start, self.cpu_count, self.counterPath, 1), options)
+        n, m = getCounters(self.counterPath)
+        self.assertEqual((n, m), (0, self.cpu_count))
+
+        resetCounters(self.counterPath)
+        Job.Runner.startToil(Job.wrapJobFn(start, self.cpu_count, self.counterPath, 2), options)
+        n, m = getCounters(self.counterPath)
+        self.assertEqual((n, m), (0, self.cpu_count / 2))
+
+    def test_static(self):
+        """
+        Tests using a PromisedRequirement to allocate number of cores in a static DAG.
+        """
+        options = Job.Runner.getDefaultOptions(self._getTestJobStorePath())
+        options.logLevel = "INFO"
+        Job.Runner.startToil(Job.wrapJobFn(static_start, self.cpu_count, self.counterPath, 1), options)
+        n, m = getCounters(self.counterPath)
+        self.assertEqual((n, m), (0, self.cpu_count))
+
+        resetCounters(self.counterPath)
+        Job.Runner.startToil(Job.wrapJobFn(static_start, self.cpu_count, self.counterPath, 2), options)
+        n, m = getCounters(self.counterPath)
+        self.assertEqual((n, m), (0, self.cpu_count / 2))
+
+
+def start(job, cpu_count, filename, scale):
+    """
+    Tests dynamic job function creation using a PromisedRequirement.
+    """
+    child = job.addChildFn(one)
+    for _ in range(cpu_count):
+        job.addFollowOnFn(increment_counter, filename,
+                          cores=PromisedRequirement(lambda x: scale * x, child.rv()),
+                          memory='1M')
+
+
+def static_start(job, cpu_count, filename, scale):
```

Comment:
Besides this being a misnomer, these two functions are nearly identical and should be refactored for DRYness.

We do need a test for static DAGs.